
<!DOCTYPE html>

<html class="no-js" lang="en">
<head>
<meta charset="utf-8"/>
<meta content="width=device-width,initial-scale=1" name="viewport"/>
<meta content="Zehua Zeng" name="author"/>
<link href="../../../img/favicon.ico" rel="icon"/>
<meta content="mkdocs-1.4.2, mkdocs-material-9.1.2" name="generator"/>
<title>Api qc - omicverse Readthedocs</title>
<link href="../../../assets/stylesheets/main.7bf56d0a.min.css" rel="stylesheet"/>
<link href="../../../assets/stylesheets/palette.a0c5b2b5.min.css" rel="stylesheet"/>
<link crossorigin="" href="https://fonts.gstatic.com" rel="preconnect"/>
<link href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&amp;display=fallback" rel="stylesheet"/>
<style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
<link href="../../../assets/_mkdocstrings.css" rel="stylesheet"/>
<link href="../../../css/custom.css" rel="stylesheet"/>
<script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
<link href="../../../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><style>
            html.glightbox-open { overflow: initial; height: 100%; }
            .gslide-title { margin-top: 0px; user-select: text; }
            .gslide-desc { color: #666; user-select: text; }
            .gslide-image img { background: white; }
            
                .gscrollbar-fixer { padding-right: 15px; }
                .gdesc-inner { font-size: 0.75rem; }
                body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color);}
                body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color);}
                body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color);}
                </style><script src="../../../assets/javascripts/glightbox.min.js"></script></head>
<body data-md-color-accent="" data-md-color-primary="" data-md-color-scheme="default" dir="ltr">
<script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
<input autocomplete="off" class="md-toggle" data-md-toggle="drawer" id="__drawer" type="checkbox"/>
<input autocomplete="off" class="md-toggle" data-md-toggle="search" id="__search" type="checkbox"/>
<label class="md-overlay" for="__drawer"></label>
<div data-md-component="skip">
<a class="md-skip" href="#omicverse.pp.qc">
          Skip to content
        </a>
</div>
<div data-md-component="announce">
</div>
<header class="md-header md-header--shadow" data-md-component="header">
<nav aria-label="Header" class="md-header__inner md-grid">
<a aria-label="omicverse Readthedocs" class="md-header__button md-logo" data-md-component="logo" href="../../.." title="omicverse Readthedocs">
<img alt="logo" src="../../../img/ico.png"/>
</a>
<label class="md-header__button md-icon" for="__drawer">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"></path></svg>
</label>
<div class="md-header__title" data-md-component="header-title">
<div class="md-header__ellipsis">
<div class="md-header__topic">
<span class="md-ellipsis">
            omicverse Readthedocs
          </span>
</div>
<div class="md-header__topic" data-md-component="header-topic">
<span class="md-ellipsis">
            
              Api qc
            
          </span>
</div>
</div>
</div>
<form class="md-header__option" data-md-component="palette">
<input aria-label="Switch to dark mode" class="md-option" data-md-color-accent="" data-md-color-media="(prefers-color-scheme: light)" data-md-color-primary="" data-md-color-scheme="default" id="__palette_1" name="__palette" type="radio"/>
<label class="md-header__button md-icon" for="__palette_2" hidden="" title="Switch to dark mode">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"></path></svg>
</label>
<input aria-label="Switch to light mode" class="md-option" data-md-color-accent="" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-primary="" data-md-color-scheme="slate" id="__palette_2" name="__palette" type="radio"/>
<label class="md-header__button md-icon" for="__palette_1" hidden="" title="Switch to light mode">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"></path></svg>
</label>
</form>
<div class="md-header__source">
<a class="md-source" data-md-component="source" href="https://github.com/Starlitnightly/omicverse" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
</nav>
</header>
<div class="md-container" data-md-component="container">
<main class="md-main" data-md-component="main">
<div class="md-main__inner md-grid">
<div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Navigation" class="md-nav md-nav--primary" data-md-level="0">
<label class="md-nav__title" for="__drawer">
<a aria-label="omicverse Readthedocs" class="md-nav__button md-logo" data-md-component="logo" href="../../.." title="omicverse Readthedocs">
<img alt="logo" src="../../../img/ico.png"/>
</a>
    omicverse Readthedocs
  </label>
<div class="md-nav__source">
<a class="md-source" data-md-component="source" href="https://github.com/Starlitnightly/omicverse" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../..">
        Index
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Installation_guild/">
        Installation
      </a>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
          Tutorial
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_label" class="md-nav" data-md-level="1">
<label class="md-nav__title" for="__nav_3">
<span class="md-nav__icon md-icon"></span>
          Tutorial
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_1" id="__nav_3_1_label" tabindex="0">
          Bulk
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_1_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_1">
<span class="md-nav__icon md-icon"></span>
          Bulk
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_bulk_combat/">
        Batch correction in Bulk RNA-seq or microarray data
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_deg/">
        Different Expression Analysis
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_deseq2/">
        Different Expression Analysis with DEseq2
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_network/">
        Protein-Protein interaction (PPI) analysis by String-db
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_wgcna/">
        WGCNA (Weighted gene co-expression network analysis) analysis
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_tcga/">
        TCGA database preprocess
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2" id="__nav_3_2_label" tabindex="0">
          Single
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_2">
<span class="md-nav__icon md-icon"></span>
          Single
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_1" id="__nav_3_2_1_label" tabindex="0">
          Preprocessing
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_1_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_1">
<span class="md-nav__icon md-icon"></span>
          Preprocessing
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_preprocess/">
        Preprocessing the data of scRNA-seq with omicverse
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cluster/">
        Clustering space
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_single_batch/">
        Data integration and batch correction
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_2" id="__nav_3_2_2_label" tabindex="0">
          Annotation
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_2_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_2">
<span class="md-nav__icon md-icon"></span>
          Annotation
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellanno/">
        Celltype auto annotation with SCSA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_metatime/">
        Celltype auto annotation with MetaTiME
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_tosica/">
        Celltype annotation migration(mapping) with TOSICA
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_3" id="__nav_3_2_3_label" tabindex="0">
          Trajectory
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_3_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_3">
<span class="md-nav__icon md-icon"></span>
          Trajectory
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_traj/">
        Trajectory Inference with PAGA or Palantir
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_via/">
        Trajectory Inference with VIA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_via_velo/">
        Trajectory Inference with VIA and scVelo
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellfate_gene/">
        Timing-associated genes analysis with cellfategenie
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellfate_genesets/">
        Timing-associated geneset analysis with cellfategenie
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_4" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_4" id="__nav_3_2_4_label" tabindex="0">
          Cell Structure
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_4_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_4">
<span class="md-nav__icon md-icon"></span>
          Cell Structure
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_aucell/">
        Pathway analysis with AUCell
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scdeg/">
        Differential expression analysis in single cell
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_simba/">
        Data integration and batch correction with SIMBA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scdrug/">
        Drug response predict with scDrug
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellphonedb/">
        Cell interaction with CellPhoneDB
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_5" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_5" id="__nav_3_2_5_label" tabindex="0">
          Multi omics
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_5_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_5">
<span class="md-nav__icon md-icon"></span>
          Multi omics
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_mofa/">
        Multi omics analysis by MOFA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_mofa_glue/">
        Multi omics analysis by MOFA and GLUE
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_anno_trans/">
        Celltype annotation transfer in multi-omics
      </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_3" id="__nav_3_3_label" tabindex="0">
          Bulk2Single
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_3_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_3">
<span class="md-nav__icon md-icon"></span>
          Bulk2Single
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_bulktrajblend/">
        Bulk RNA-seq generate 'interrupted' cells to interpolate scRNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_bulk2single/">
        Bulk RNA-seq to Single RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_single2spatial/">
        Single RNA-seq to Spatial RNA-seq
      </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../">
        API documentation
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Release_notes/">
        Release notes
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Developer_guild/">
        Developer guild
      </a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Table of contents" class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">
<span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
<ul class="md-nav__list" data-md-component="toc" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.pp.qc">
    qc()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.utils.store_layers">
    store_layers()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.utils.retrieve_layers">
    retrieve_layers()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.utils.anndata_sparse">
    anndata_sparse()
  </a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-content" data-md-component="content">
<article class="md-content__inner md-typeset">
<h1>Api qc</h1>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.pp.qc">
<code class="highlight language-python"><span class="n">omicverse</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">qc</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">'seurat'</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">nmads</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">batch_key</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">path_viz</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">tresh</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.pp.qc" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Perform quality control on a dictionary of AnnData objects.</p>
<p>This function calculates several QC metrics, including mitochondrial percentage, nUMIs, 
and detected genes, and produces several plots visualizing the QC metrics for each sample. 
The function performs doublet detection using scrublet and filtering using either 
Seurat or MADs. The function returns a merged AnnData object with cells that passed QC filters
and a list of cells that did not pass QC on all samples.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adatas</code></td>
<td>
</td>
<td><p>AnnData objects</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>mode</code></td>
<td>
</td>
<td><p>The filtering method to use. Valid options are 'seurat' and 'mads'. Default is 'seurat'.</p></td>
<td>
<code>'seurat'</code>
</td>
</tr>
<tr>
<td><code>min_cells</code></td>
<td>
</td>
<td><p>The minimum number of cells for a sample to pass QC. Default is 3.</p></td>
<td>
<code>3</code>
</td>
</tr>
<tr>
<td><code>min_genes</code></td>
<td>
</td>
<td><p>The minimum number of genes for a cell to pass QC. Default is 200.</p></td>
<td>
<code>200</code>
</td>
</tr>
<tr>
<td><code>nmads</code></td>
<td>
</td>
<td><p>The number of MADs to use for MADs filtering. Default is 5.</p></td>
<td>
<code>5</code>
</td>
</tr>
<tr>
<td><code>path_viz</code></td>
<td>
</td>
<td><p>The path to save the QC plots. Default is None.</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>tresh</code></td>
<td>
</td>
<td><p>A dictionary of QC thresholds. The keys should be 'mito_perc', 'nUMIs', and 'detected_genes'.
Only used if mode is 'seurat'. Default is None.</p></td>
<td>
<code>None</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td> <td>
</td>
<td><p>An AnnData object containing cells that passed QC filters.</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/pp/_qc.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="176 "></span><span class="k">def</span> <span class="nf">qc</span><span class="p">(</span><span class="n">adata</span><span class="p">:</span><span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">'seurat'</span><span class="p">,</span> 
<span class="linenos" data-linenos="177 "></span>       <span class="n">min_cells</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">nmads</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> 
<span class="linenos" data-linenos="178 "></span>       <span class="n">batch_key</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="linenos" data-linenos="179 "></span>       <span class="n">path_viz</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">tresh</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="linenos" data-linenos="180 "></span><span class="w">    </span><span class="sd">"""</span>
<span class="linenos" data-linenos="181 "></span><span class="sd">    Perform quality control on a dictionary of AnnData objects.</span>
<span class="linenos" data-linenos="182 "></span>
<span class="linenos" data-linenos="183 "></span><span class="sd">    This function calculates several QC metrics, including mitochondrial percentage, nUMIs, </span>
<span class="linenos" data-linenos="184 "></span><span class="sd">    and detected genes, and produces several plots visualizing the QC metrics for each sample. </span>
<span class="linenos" data-linenos="185 "></span><span class="sd">    The function performs doublet detection using scrublet and filtering using either </span>
<span class="linenos" data-linenos="186 "></span><span class="sd">    Seurat or MADs. The function returns a merged AnnData object with cells that passed QC filters</span>
<span class="linenos" data-linenos="187 "></span><span class="sd">    and a list of cells that did not pass QC on all samples.</span>
<span class="linenos" data-linenos="188 "></span>
<span class="linenos" data-linenos="189 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="190 "></span><span class="sd">        adatas : AnnData objects</span>
<span class="linenos" data-linenos="191 "></span><span class="sd">        mode : The filtering method to use. Valid options are 'seurat' and 'mads'. Default is 'seurat'.</span>
<span class="linenos" data-linenos="192 "></span><span class="sd">        min_cells : The minimum number of cells for a sample to pass QC. Default is 3.</span>
<span class="linenos" data-linenos="193 "></span><span class="sd">        min_genes : The minimum number of genes for a cell to pass QC. Default is 200.</span>
<span class="linenos" data-linenos="194 "></span><span class="sd">        nmads : The number of MADs to use for MADs filtering. Default is 5.</span>
<span class="linenos" data-linenos="195 "></span><span class="sd">        path_viz : The path to save the QC plots. Default is None.</span>
<span class="linenos" data-linenos="196 "></span><span class="sd">        tresh : A dictionary of QC thresholds. The keys should be 'mito_perc', 'nUMIs', and 'detected_genes'.</span>
<span class="linenos" data-linenos="197 "></span><span class="sd">            Only used if mode is 'seurat'. Default is None.</span>
<span class="linenos" data-linenos="198 "></span>
<span class="linenos" data-linenos="199 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="200 "></span><span class="sd">        adata : An AnnData object containing cells that passed QC filters.</span>
<span class="linenos" data-linenos="201 "></span>
<span class="linenos" data-linenos="202 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="203 "></span>    <span class="c1"># Logging </span>
<span class="linenos" data-linenos="204 "></span>    <span class="k">if</span> <span class="n">tresh</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<span class="linenos" data-linenos="205 "></span>        <span class="n">tresh</span><span class="o">=</span><span class="p">{</span><span class="s1">'mito_perc'</span><span class="p">:</span> <span class="mf">0.15</span><span class="p">,</span> <span class="s1">'nUMIs'</span><span class="p">:</span> <span class="mi">500</span><span class="p">,</span> <span class="s1">'detected_genes'</span><span class="p">:</span> <span class="mi">250</span><span class="p">}</span>
<span class="linenos" data-linenos="206 "></span>
<span class="linenos" data-linenos="207 "></span>    <span class="c1"># For each adata, produce a figure</span>
<span class="linenos" data-linenos="208 "></span>    <span class="c1"># with PdfPages(path_viz + 'original_QC_by_sample.pdf') as pdf: </span>
<span class="linenos" data-linenos="209 "></span>    <span class="n">removed_cells</span> <span class="o">=</span> <span class="p">[]</span>
<span class="linenos" data-linenos="210 "></span>
<span class="linenos" data-linenos="211 "></span>    <span class="c1"># QC metrics</span>
<span class="linenos" data-linenos="212 "></span>    <span class="nb">print</span><span class="p">(</span><span class="s1">'Calculate QC metrics'</span><span class="p">)</span>
<span class="linenos" data-linenos="213 "></span>    <span class="n">adata</span><span class="o">.</span><span class="n">var_names_make_unique</span><span class="p">()</span>
<span class="linenos" data-linenos="214 "></span>    <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">"mt"</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">"MT-"</span><span class="p">)</span>
<span class="linenos" data-linenos="215 "></span>    <span class="k">if</span> <span class="n">issparse</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">):</span>
<span class="linenos" data-linenos="216 "></span>        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'nUMIs'</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>  
<span class="linenos" data-linenos="217 "></span>        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'mito_perc'</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">"mt"</span><span class="p">]]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'nUMIs'</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="linenos" data-linenos="218 "></span>        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'detected_genes'</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>  
<span class="linenos" data-linenos="219 "></span>    <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="220 "></span>        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'nUMIs'</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>  
<span class="linenos" data-linenos="221 "></span>        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'mito_perc'</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">"mt"</span><span class="p">]]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'nUMIs'</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="linenos" data-linenos="222 "></span>        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'detected_genes'</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>  
<span class="linenos" data-linenos="223 "></span>    <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'cell_complexity'</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'detected_genes'</span><span class="p">]</span> <span class="o">/</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'nUMIs'</span><span class="p">]</span>
<span class="linenos" data-linenos="224 "></span>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">'End calculation of QC metrics.'</span><span class="p">)</span>
<span class="linenos" data-linenos="225 "></span>
<span class="linenos" data-linenos="226 "></span>    <span class="c1"># Original QC plot</span>
<span class="linenos" data-linenos="227 "></span>    <span class="n">n0</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="linenos" data-linenos="228 "></span>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">'Original cell number: </span><span class="si">{</span><span class="n">n0</span><span class="si">}</span><span class="s1">'</span><span class="p">)</span>
<span class="linenos" data-linenos="229 "></span>
<span class="linenos" data-linenos="230 "></span>    <span class="c1"># Post doublets removal QC plot</span>
<span class="linenos" data-linenos="231 "></span>    <span class="nb">print</span><span class="p">(</span><span class="s1">'Begin of post doublets removal and QC plot'</span><span class="p">)</span>
<span class="linenos" data-linenos="232 "></span>    <span class="n">sc</span><span class="o">.</span><span class="n">external</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">scrublet</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">1234</span><span class="p">,</span><span class="n">batch_key</span><span class="o">=</span><span class="n">batch_key</span><span class="p">)</span>
<span class="linenos" data-linenos="233 "></span>    <span class="n">adata_remove</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'predicted_doublet'</span><span class="p">],</span> <span class="p">:]</span>
<span class="linenos" data-linenos="234 "></span>    <span class="n">removed_cells</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">adata_remove</span><span class="o">.</span><span class="n">obs_names</span><span class="p">))</span>
<span class="linenos" data-linenos="235 "></span>    <span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="o">~</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'predicted_doublet'</span><span class="p">],</span> <span class="p">:]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="linenos" data-linenos="236 "></span>    <span class="n">n1</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="linenos" data-linenos="237 "></span>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">'Cells retained after scrublet: </span><span class="si">{</span><span class="n">n1</span><span class="si">}</span><span class="s1">, </span><span class="si">{</span><span class="n">n0</span><span class="o">-</span><span class="n">n1</span><span class="si">}</span><span class="s1"> removed.'</span><span class="p">)</span>
<span class="linenos" data-linenos="238 "></span>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">'End of post doublets removal and QC plots.'</span><span class="p">)</span>
<span class="linenos" data-linenos="239 "></span>
<span class="linenos" data-linenos="240 "></span>    <span class="c1"># Post seurat or mads filtering QC plot</span>
<span class="linenos" data-linenos="241 "></span>
<span class="linenos" data-linenos="242 "></span>    <span class="c1"># Filters</span>
<span class="linenos" data-linenos="243 "></span>    <span class="nb">print</span><span class="p">(</span><span class="s1">'Filters application (seurat or mads)'</span><span class="p">)</span>
<span class="linenos" data-linenos="244 "></span>    <span class="k">if</span> <span class="n">mode</span> <span class="o">==</span> <span class="s1">'seurat'</span><span class="p">:</span>
<span class="linenos" data-linenos="245 "></span>        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'passing_mt'</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'mito_perc'</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">tresh</span><span class="p">[</span><span class="s1">'mito_perc'</span><span class="p">]</span>
<span class="linenos" data-linenos="246 "></span>        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'passing_nUMIs'</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'nUMIs'</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">tresh</span><span class="p">[</span><span class="s1">'nUMIs'</span><span class="p">]</span>
<span class="linenos" data-linenos="247 "></span>        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'passing_ngenes'</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'detected_genes'</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">tresh</span><span class="p">[</span><span class="s1">'detected_genes'</span><span class="p">]</span>
<span class="linenos" data-linenos="248 "></span>    <span class="k">elif</span> <span class="n">mode</span> <span class="o">==</span> <span class="s1">'mads'</span><span class="p">:</span>
<span class="linenos" data-linenos="249 "></span>        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'passing_mt'</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'mito_perc'</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">tresh</span><span class="p">[</span><span class="s1">'mito_perc'</span><span class="p">]</span>
<span class="linenos" data-linenos="250 "></span>        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'passing_nUMIs'</span><span class="p">]</span> <span class="o">=</span> <span class="n">mads_test</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="s1">'nUMIs'</span><span class="p">,</span> <span class="n">nmads</span><span class="o">=</span><span class="n">nmads</span><span class="p">,</span> <span class="n">lt</span><span class="o">=</span><span class="n">tresh</span><span class="p">)</span>
<span class="linenos" data-linenos="251 "></span>        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'passing_ngenes'</span><span class="p">]</span> <span class="o">=</span> <span class="n">mads_test</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="s1">'detected_genes'</span><span class="p">,</span> <span class="n">nmads</span><span class="o">=</span><span class="n">nmads</span><span class="p">,</span> <span class="n">lt</span><span class="o">=</span><span class="n">tresh</span><span class="p">)</span>  
<span class="linenos" data-linenos="252 "></span>
<span class="linenos" data-linenos="253 "></span>    <span class="c1"># Report </span>
<span class="linenos" data-linenos="254 "></span>    <span class="k">if</span> <span class="n">mode</span> <span class="o">==</span> <span class="s1">'seurat'</span><span class="p">:</span>
<span class="linenos" data-linenos="255 "></span>        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">'Lower treshold, nUMIs: </span><span class="si">{</span><span class="n">tresh</span><span class="p">[</span><span class="s2">"nUMIs"</span><span class="p">]</span><span class="si">}</span><span class="s1">; filtered-out-cells: </span><span class="si">{</span><span class="n">n1</span><span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">"passing_nUMIs"</span><span class="p">])</span><span class="si">}</span><span class="s1">'</span><span class="p">)</span>
<span class="linenos" data-linenos="256 "></span>        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">'Lower treshold, n genes: </span><span class="si">{</span><span class="n">tresh</span><span class="p">[</span><span class="s2">"detected_genes"</span><span class="p">]</span><span class="si">}</span><span class="s1">; filtered-out-cells: </span><span class="si">{</span><span class="n">n1</span><span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">"passing_ngenes"</span><span class="p">])</span><span class="si">}</span><span class="s1">'</span><span class="p">)</span>
<span class="linenos" data-linenos="257 "></span>        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">'Lower treshold, mito %: </span><span class="si">{</span><span class="n">tresh</span><span class="p">[</span><span class="s2">"mito_perc"</span><span class="p">]</span><span class="si">}</span><span class="s1">; filtered-out-cells: </span><span class="si">{</span><span class="n">n1</span><span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">"passing_mt"</span><span class="p">])</span><span class="si">}</span><span class="s1">'</span><span class="p">)</span>
<span class="linenos" data-linenos="258 "></span>    <span class="k">elif</span> <span class="n">mode</span> <span class="o">==</span> <span class="s1">'mads'</span><span class="p">:</span>
<span class="linenos" data-linenos="259 "></span>        <span class="n">nUMIs_t</span> <span class="o">=</span> <span class="n">mads</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="s1">'nUMIs'</span><span class="p">,</span> <span class="n">nmads</span><span class="o">=</span><span class="n">nmads</span><span class="p">,</span> <span class="n">lt</span><span class="o">=</span><span class="n">tresh</span><span class="p">)</span>
<span class="linenos" data-linenos="260 "></span>        <span class="n">n_genes_t</span> <span class="o">=</span> <span class="n">mads</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="s1">'detected_genes'</span><span class="p">,</span> <span class="n">nmads</span><span class="o">=</span><span class="n">nmads</span><span class="p">,</span> <span class="n">lt</span><span class="o">=</span><span class="n">tresh</span><span class="p">)</span>
<span class="linenos" data-linenos="261 "></span>        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">'Tresholds used, nUMIs: (</span><span class="si">{</span><span class="n">nUMIs_t</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">, </span><span class="si">{</span><span class="n">nUMIs_t</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s1">); filtered-out-cells: </span><span class="si">{</span><span class="n">n1</span><span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">"passing_nUMIs"</span><span class="p">])</span><span class="si">}</span><span class="s1">'</span><span class="p">)</span>
<span class="linenos" data-linenos="262 "></span>        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">'Tresholds used, n genes: (</span><span class="si">{</span><span class="n">n_genes_t</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">, </span><span class="si">{</span><span class="n">n_genes_t</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s1">); filtered-out-cells: </span><span class="si">{</span><span class="n">n1</span><span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">"passing_ngenes"</span><span class="p">])</span><span class="si">}</span><span class="s1">'</span><span class="p">)</span>
<span class="linenos" data-linenos="263 "></span>        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">'Lower treshold, mito %: </span><span class="si">{</span><span class="n">tresh</span><span class="p">[</span><span class="s2">"mito_perc"</span><span class="p">]</span><span class="si">}</span><span class="s1">; filtered-out-cells: </span><span class="si">{</span><span class="n">n1</span><span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">"passing_mt"</span><span class="p">])</span><span class="si">}</span><span class="s1">'</span><span class="p">)</span>
<span class="linenos" data-linenos="264 "></span>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">'Filters applicated.'</span><span class="p">)</span>
<span class="linenos" data-linenos="265 "></span>
<span class="linenos" data-linenos="266 "></span>    <span class="c1"># QC plot</span>
<span class="linenos" data-linenos="267 "></span>    <span class="n">QC_test</span> <span class="o">=</span> <span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'passing_mt'</span><span class="p">])</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'passing_nUMIs'</span><span class="p">])</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'passing_ngenes'</span><span class="p">])</span>
<span class="linenos" data-linenos="268 "></span>    <span class="n">removed</span> <span class="o">=</span> <span class="n">QC_test</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="k">lambda</span> <span class="n">x</span> <span class="p">:</span> <span class="n">x</span> <span class="o">==</span> <span class="kc">False</span><span class="p">]</span>
<span class="linenos" data-linenos="269 "></span>    <span class="n">removed_cells</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">removed</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="p">))</span>
<span class="linenos" data-linenos="270 "></span>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">'Total cell filtered out with this last --mode </span><span class="si">{</span><span class="n">mode</span><span class="si">}</span><span class="s1"> QC (and its chosen options): </span><span class="si">{</span><span class="n">n1</span><span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">QC_test</span><span class="p">)</span><span class="si">}</span><span class="s1">'</span><span class="p">)</span>
<span class="linenos" data-linenos="271 "></span>    <span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">QC_test</span><span class="p">,</span> <span class="p">:]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="linenos" data-linenos="272 "></span>    <span class="n">n2</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="linenos" data-linenos="273 "></span>
<span class="linenos" data-linenos="274 "></span>
<span class="linenos" data-linenos="275 "></span>
<span class="linenos" data-linenos="276 "></span>    <span class="c1"># Store cleaned adata</span>
<span class="linenos" data-linenos="277 "></span>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">'Cells retained after scrublet and </span><span class="si">{</span><span class="n">mode</span><span class="si">}</span><span class="s1"> filtering: </span><span class="si">{</span><span class="n">n2</span><span class="si">}</span><span class="s1">, </span><span class="si">{</span><span class="n">n0</span><span class="o">-</span><span class="n">n2</span><span class="si">}</span><span class="s1"> removed.'</span><span class="p">)</span>
<span class="linenos" data-linenos="278 "></span>
<span class="linenos" data-linenos="279 "></span>    <span class="c1"># Last gene and cell filter</span>
<span class="linenos" data-linenos="280 "></span>    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="n">min_genes</span><span class="p">)</span>
<span class="linenos" data-linenos="281 "></span>    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="n">min_cells</span><span class="p">)</span>
<span class="linenos" data-linenos="282 "></span>
<span class="linenos" data-linenos="283 "></span>    <span class="k">return</span> <span class="n">adata</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.utils.store_layers">
<code class="highlight language-python"><span class="n">omicverse</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">store_layers</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">layers</span><span class="o">=</span><span class="s1">'counts'</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.utils.store_layers" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Store the X of adata in adata.uns['layers_{}'.format(layers)]</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td>
<td>
</td>
<td><p>AnnData</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>layers</code></td>
<td>
</td>
<td><p>the layers name to store, default 'counts'</p></td>
<td>
<code>'counts'</code>
</td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/utils/_data.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="535 "></span><span class="k">def</span> <span class="nf">store_layers</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">layers</span><span class="o">=</span><span class="s1">'counts'</span><span class="p">):</span>
<span class="linenos" data-linenos="536 "></span><span class="w">    </span><span class="sd">"""</span>
<span class="linenos" data-linenos="537 "></span><span class="sd">    Store the X of adata in adata.uns['layers_{}'.format(layers)]</span>
<span class="linenos" data-linenos="538 "></span>
<span class="linenos" data-linenos="539 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="540 "></span><span class="sd">        adata: AnnData</span>
<span class="linenos" data-linenos="541 "></span><span class="sd">        layers: the layers name to store, default 'counts'</span>
<span class="linenos" data-linenos="542 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="543 "></span>
<span class="linenos" data-linenos="544 "></span>
<span class="linenos" data-linenos="545 "></span>    <span class="k">if</span> <span class="n">issparse</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">isspmatrix_csr</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">):</span>
<span class="linenos" data-linenos="546 "></span>        <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">'layers_</span><span class="si">{}</span><span class="s1">'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">layers</span><span class="p">)]</span><span class="o">=</span><span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">csr_matrix</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()),</span>
<span class="linenos" data-linenos="547 "></span>                                           <span class="n">obs</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">),</span>
<span class="linenos" data-linenos="548 "></span>                                          <span class="n">var</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">),)</span>
<span class="linenos" data-linenos="549 "></span>    <span class="k">elif</span> <span class="n">issparse</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">):</span>
<span class="linenos" data-linenos="550 "></span>        <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">'layers_</span><span class="si">{}</span><span class="s1">'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">layers</span><span class="p">)]</span><span class="o">=</span><span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">(),</span>
<span class="linenos" data-linenos="551 "></span>                                           <span class="n">obs</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">),</span>
<span class="linenos" data-linenos="552 "></span>                                           <span class="n">var</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">),)</span>
<span class="linenos" data-linenos="553 "></span>    <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="554 "></span>        <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">'layers_</span><span class="si">{}</span><span class="s1">'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">layers</span><span class="p">)]</span><span class="o">=</span><span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">csr_matrix</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()),</span>
<span class="linenos" data-linenos="555 "></span>                                           <span class="n">obs</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">),</span>
<span class="linenos" data-linenos="556 "></span>                                          <span class="n">var</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">),)</span>
<span class="linenos" data-linenos="557 "></span>    <span class="nb">print</span><span class="p">(</span><span class="s1">'......The X of adata have been stored in </span><span class="si">{}</span><span class="s1">'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">layers</span><span class="p">))</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.utils.retrieve_layers">
<code class="highlight language-python"><span class="n">omicverse</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">retrieve_layers</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">layers</span><span class="o">=</span><span class="s1">'counts'</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.utils.retrieve_layers" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Retrieve the X of adata from adata.uns['layers_{}'.format(layers)]</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td>
<td>
</td>
<td><p>AnnData</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>layers</code></td>
<td>
</td>
<td><p>the layers name to retrieve, default 'counts'</p></td>
<td>
<code>'counts'</code>
</td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/utils/_data.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="559 "></span><span class="k">def</span> <span class="nf">retrieve_layers</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">layers</span><span class="o">=</span><span class="s1">'counts'</span><span class="p">):</span>
<span class="linenos" data-linenos="560 "></span><span class="w">    </span><span class="sd">"""</span>
<span class="linenos" data-linenos="561 "></span><span class="sd">    Retrieve the X of adata from adata.uns['layers_{}'.format(layers)]</span>
<span class="linenos" data-linenos="562 "></span>
<span class="linenos" data-linenos="563 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="564 "></span><span class="sd">        adata: AnnData</span>
<span class="linenos" data-linenos="565 "></span><span class="sd">        layers: the layers name to retrieve, default 'counts'</span>
<span class="linenos" data-linenos="566 "></span>
<span class="linenos" data-linenos="567 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="568 "></span>
<span class="linenos" data-linenos="569 "></span>    <span class="n">adata_test</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">'layers_</span><span class="si">{}</span><span class="s1">'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">layers</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="linenos" data-linenos="570 "></span>    <span class="n">adata_test</span><span class="o">=</span><span class="n">adata_test</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">,</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
<span class="linenos" data-linenos="571 "></span>
<span class="linenos" data-linenos="572 "></span>    <span class="k">if</span> <span class="n">issparse</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">isspmatrix_csr</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">):</span>
<span class="linenos" data-linenos="573 "></span>        <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">'layers_raw'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">layers</span><span class="p">)]</span><span class="o">=</span><span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">csr_matrix</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()),</span>
<span class="linenos" data-linenos="574 "></span>                                           <span class="n">obs</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">),</span>
<span class="linenos" data-linenos="575 "></span>                                          <span class="n">var</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">),)</span>
<span class="linenos" data-linenos="576 "></span>    <span class="k">elif</span> <span class="n">issparse</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">):</span>
<span class="linenos" data-linenos="577 "></span>        <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">'layers_raw'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">layers</span><span class="p">)]</span><span class="o">=</span><span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">(),</span>
<span class="linenos" data-linenos="578 "></span>                                           <span class="n">obs</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">),</span>
<span class="linenos" data-linenos="579 "></span>                                           <span class="n">var</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">),)</span>
<span class="linenos" data-linenos="580 "></span>    <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="581 "></span>        <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">'layers_raw'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">layers</span><span class="p">)]</span><span class="o">=</span><span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">csr_matrix</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()),</span>
<span class="linenos" data-linenos="582 "></span>                                           <span class="n">obs</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">),</span>
<span class="linenos" data-linenos="583 "></span>                                          <span class="n">var</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">),)</span>
<span class="linenos" data-linenos="584 "></span>    <span class="nb">print</span><span class="p">(</span><span class="s1">'......The X of adata have been stored in raw'</span><span class="p">)</span>
<span class="linenos" data-linenos="585 "></span>    <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">=</span><span class="n">adata_test</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="linenos" data-linenos="586 "></span>    <span class="nb">print</span><span class="p">(</span><span class="s1">'......The layers </span><span class="si">{}</span><span class="s1"> of adata have been retreved'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">layers</span><span class="p">))</span>
<span class="linenos" data-linenos="587 "></span>    <span class="k">del</span> <span class="n">adata_test</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.utils.anndata_sparse">
<code class="highlight language-python"><span class="n">omicverse</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">anndata_sparse</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.utils.anndata_sparse" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Set adata.X to csr_matrix</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td>
<td>
</td>
<td><p>AnnData</p></td>
<td>
<em>required</em>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td> <td>
</td>
<td><p>AnnData</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/utils/_data.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="518 "></span><span class="k">def</span> <span class="nf">anndata_sparse</span><span class="p">(</span><span class="n">adata</span><span class="p">):</span>
<span class="linenos" data-linenos="519 "></span><span class="w">    </span><span class="sd">"""</span>
<span class="linenos" data-linenos="520 "></span><span class="sd">    Set adata.X to csr_matrix</span>
<span class="linenos" data-linenos="521 "></span>
<span class="linenos" data-linenos="522 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="523 "></span><span class="sd">        adata: AnnData</span>
<span class="linenos" data-linenos="524 "></span>
<span class="linenos" data-linenos="525 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="526 "></span><span class="sd">        adata: AnnData</span>
<span class="linenos" data-linenos="527 "></span>
<span class="linenos" data-linenos="528 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="529 "></span>
<span class="linenos" data-linenos="530 "></span>    <span class="kn">from</span> <span class="nn">scipy.sparse</span> <span class="kn">import</span> <span class="n">csr_matrix</span>
<span class="linenos" data-linenos="531 "></span>    <span class="n">x</span> <span class="o">=</span> <span class="n">csr_matrix</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">())</span>
<span class="linenos" data-linenos="532 "></span>    <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">=</span><span class="n">x</span>
<span class="linenos" data-linenos="533 "></span>    <span class="k">return</span> <span class="n">adata</span>
</code></pre></div>
</details>
</div>
</div>
</article>
</div>
</div>
</main>
<footer class="md-footer">
<div class="md-footer-meta md-typeset">
<div class="md-footer-meta__inner md-grid">
<div class="md-copyright">
<div class="md-copyright__highlight">
      Copyright © 2019-2023, 112 Lab, USTB
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" rel="noopener" target="_blank">
      Material for MkDocs
    </a>
</div>
</div>
</div>
</footer>
</div>
<div class="md-dialog" data-md-component="dialog">
<div class="md-dialog__inner md-typeset"></div>
</div>
<script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.instant", "navigation.tracking", "navigation.indexes"], "search": "../../../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
<script src="../../../assets/javascripts/bundle.fc8c2696.min.js"></script>
<script src="../../../javascripts/config.js"></script>
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
<script>document$.subscribe(() => {const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});})</script></body>
</html>